src/main/scala/beam/agentsim/infrastructure/ChargingNetworkManager.scala

```diff
@@ -0,0 +1,477 @@
+      val chargingNetwork = chargingNetworkMap(vehicleManager)
+      chargingNetwork.lookupVehicle(vehicle.id) match {
+        case Some(chargingVehicle) =>
+          val prevStartTime = chargingVehicle.latestChargingCycle.get.startTime
+          val startTime = Math.min(tick, prevStartTime)
+          val endTime = Math.max(tick, prevStartTime)
+          val duration = endTime - startTime
+          val (chargeDurationAtTick, energyToChargeAtTick) = dispatchEnergy(duration, chargingVehicle, physicalBounds)
+          chargingVehicle.processChargingCycle(startTime, energyToChargeAtTick, chargeDurationAtTick)
+          handleEndCharging(tick, chargingVehicle, Some(sender))
+        case _ =>
+          log.debug(s"Vehicle $vehicle is already disconnected at $tick")
+          sender ! UnhandledVehicle(tick, vehicle.id)
+      }
+
+    case Finish =>
+      chargingNetworkMap.foreach(_._2.clear())
+      chargingNetworkMap.clear()
+      powerController.close()
+      context.children.foreach(_ ! Finish)
+      context.stop(self)
+  }
+
+  /**
+    * Connect the vehicle
+    * @param tick current time
+    * @param chargingVehicle charging vehicle information
+    */
+  private def handleStartCharging(tick: Int, chargingVehicle: ChargingVehicle): Unit = {
+    val ChargingVehicle(vehicle, vehicleManager, _, _, _, theSender) = chargingVehicle
+    log.debug(s"Starting charging for vehicle $vehicle at $tick")
+    val physicalBounds = obtainPowerPhysicalBounds(tick, None)
+    vehicle.connectToChargingPoint(tick)
+    theSender ! StartingRefuelSession(tick, vehicle.id)
+    handleStartChargingHelper(tick, chargingVehicle)
+    val (chargeDurationAtTick, energyToChargeAtTick) =
+      dispatchEnergy(nextTimeBin(tick) - tick, chargingVehicle, physicalBounds)
+    chargingVehicle.processChargingCycle(tick, energyToChargeAtTick, chargeDurationAtTick)
+    val endTime = chargingVehicle.computeSessionEndTime
+    if (endTime < nextTimeBin(tick))
+      scheduler ! ScheduleTrigger(ChargingTimeOutTrigger(endTime, vehicle.id, vehicleManager), self)
+  }
+
+  /**
+    * Disconnect the vehicle
+    * @param tick current time
+    * @param chargingVehicle charging vehicle information
+    * @param currentSenderMaybe the current actor who sent the message
+    */
+  private def handleEndCharging(
+    tick: Int,
+    chargingVehicle: ChargingVehicle,
+    currentSenderMaybe: Option[ActorRef] = None
+  ): Unit = {
+    val ChargingVehicle(vehicle, vehicleManager, _, station, _, _) = chargingVehicle
+    val chargingNetwork = chargingNetworkMap(vehicleManager)
+    if (chargingNetwork.disconnectVehicle(vehicle)) {
+      handleRefueling(chargingVehicle)
+      handleEndChargingHelper(tick, chargingVehicle)
+      vehicle.disconnectFromChargingPoint()
+      vehicle.stall match {
+        case Some(stall) =>
+          parkingManager ! ReleaseParkingStall(stall.parkingZoneId, stall.tazId)
+          vehicle.unsetParkingStall()
+        case None =>
```

Comment:
As above - if this happens then the disconnectVehicle will throw and it will never reach this point anyway.

Reply:
yes. But I remember there was some concurrency issue around it. I will remove that and see what happens